components/tidb_query/src/batch/executors/top_n_executor.rs

```diff
@@ -388,19 +414,30 @@ impl Ord for HeapItemUnsafe {
 
             // There is panic inside, but will never panic, since the data type of corresponding
             // column should be consistent for each `HeapItemUnsafe`.
-            let ord = lhs.cmp(&rhs);
+            let ord = lhs.cmp_with_collation(
+                &rhs,
+                order_exprs_field_type[column_idx].as_accessor().collation(),
+            )?;
 
             if ord == Ordering::Equal {
                 continue;
             }
             if !order_is_desc[column_idx] {
-                return ord;
+                return Ok(ord);
             } else {
-                return ord.reverse();
+                return Ok(ord.reverse());
             }
         }
 
-        Ordering::Equal
+        Ok(Ordering::Equal)
+    }
+}
+
+/// WARN: HeapItemUnsafe implements partial ordering. It panics when Collator fails to parse.
```

Comment:
I don't fully understand the comment `HeapItemUnsafe implements partial ordering.`. Do you mean that it only satisfies partial order in principle but in fact it must satisfy total order?

Reply:
What I mean is that it's not comparable in some states, e.g not valid utf8 byte slice, which is the definition of partial ordering. Maybe it would be less confusing if I remove this statement.

Reply:
The `Ord` is required by `BinaryHeap`, so the check is necessary to ensure that all data fed to `BinaryHeap` is the total ordering part in set.